rfd/0010-api.md

```diff
@@ -35,89 +34,211 @@ func connectClient() (*auth.Client, error) {
+
+### api module
+
+The api package will be made a sub module of teleport. It will exist in the `teleport/api`, and will be imported by the main `teleport` module through a go.mod [replace directive](https://golang.org/ref/mod#go-mod-file-replace) using a relative path.
+
+Users will be able to get the latest api version by using `go get github.com/gravitational/teleport/api`, or a specific version with `go get github.com/gravitational/teleport/api@vX.Y.Z`.
+
+#### versioning
+
+Go module versioning must follow semver and can automatically be pushed with git tags. Since the api module is a sub module, its version tags will look like `git tag api/vX.Y.Z` (this is [required](https://github.com/golang/go/wiki/Modules#is-it-possible-to-add-a-module-to-a-multi-module-repository) for multi-module repositories).
+
+If the module's version is above 2, it must follow the [v2+ Go Module release guide](https://github.com/golang/go/wiki/Modules#releasing-modules-v2-or-higher) by appending a major version suffix to the import path. The import path for each major version `X` will look like `github.com/gravitational/teleport/api/vX`.
+
+Note that if these go module requirements aren't implemented, the api module will be versioned as v0.0.0-[random-string]. Getting the latest version with `go get github.com/gravitational/teleport/api` will get the most recent version in terms of time rather than semver/teleport version. For this reason it is important to release a proper public version as soon as possible.
+
+##### Compatibility
+
+The gRPC API and other code imported by `github.com/gravitational/teleport` must follow the [Teleport version compatibility](https://github.com/gravitational/teleport/blob/master/rfd/0012-teleport-versioning.md#compatibility) rules. However, the api module itself does not need to follow the same versioning rules. For example, if the API module has a new feature that is not used by the main teleport module, such as the Credential Loaders, it does not need to be compatible with previous versions of Teleport.
+
+This gives us some flexibility to choose an api versioning strategy without worries of unrelated versioning promises.
+
+##### Release option 1 - Release the API with each Teleport release
+
+The API can be tagged for release as part of the existing Teleport release process. Each Teleport release will get both the `vX.Y.Z` and `api/vX.Y.Z` tags.
+
+This can be done within the current [release checklist](https://gravitational.slab.com/posts/teleport-release-checklist-cf0s6vmt) with just a few additional steps:
+ - Add git tag - `api/vX.Y.Z`
+ - If it's a major version bump...
+   - Update `api/go.mod` with the correct `vX` major version suffix.
+   - Update `teleport/go.mod` to import the new version suffix and update any `teleport/api` imports with the updated suffix.
```

Comment:
We should ideally automate these steps as much as possible to make sure that they aren't missed from a release.